core/block_validator.go

```diff
@@ -0,0 +1,228 @@
+func NewBlockValidator(blockchain *BlockChain, pow pow.PoW) *BlockValidator {
+	sm := &BlockValidator{
+		Pow: pow,
+		bc:  blockchain,
+	}
+	return sm
+}
+
+func (v *BlockValidator) ValidateBlock(block *types.Block) error {
+	if v.bc.HasBlock(block.Hash()) {
+		return &KnownBlockError{block.Number(), block.Hash()}
+	}
+	if !v.bc.HasBlock(block.ParentHash()) {
+		return ParentError(block.ParentHash())
+	}
+	var (
+		parent = v.bc.GetBlock(block.ParentHash())
+		header = block.Header()
+	)
+	// validate the block header
+	if err := ValidateHeader(v.Pow, header, parent.Header(), false, false); err != nil {
+		return err
+	}
+	// varify the uncles are correctly rewarded
+	if err := v.VerifyUncles(block, parent); err != nil {
+		return err
+	}
+
+	// Verify UncleHash before running other uncle validations
+	unclesSha := types.CalcUncleHash(block.Uncles())
+	if unclesSha != header.UncleHash {
+		return fmt.Errorf("invalid uncles root hash. received=%x calculated=%x", header.UncleHash, unclesSha)
+	}
+
+	// The transactions Trie's root (R = (Tr [[i, RLP(T1)], [i, RLP(T2)], ... [n, RLP(Tn)]]))
+	// can be used by light clients to make sure they've received the correct Txs
+	txSha := types.DeriveSha(block.Transactions())
+	if txSha != header.TxHash {
+		return fmt.Errorf("invalid transaction root hash. received=%x calculated=%x", header.TxHash, txSha)
+	}
+
+	return nil
+}
+
+func (v *BlockValidator) ValidateState(block, parent *types.Block, state *state.StateDB, receipts types.Receipts, usedGas *big.Int) (batch ethdb.Batch, err error) {
+	var (
+		header = block.Header()
+	)
+	if block.GasUsed().Cmp(usedGas) != 0 {
+		err = ValidationError(fmt.Sprintf("gas used error (%v / %v)", block.GasUsed(), usedGas))
+		return
+	}
+	// Validate the received block's bloom with the one derived from the generated receipts.
+	// For valid blocks this should always validate to true.
+	rbloom := types.CreateBloom(receipts)
+	if rbloom != header.Bloom {
+		err = fmt.Errorf("unable to replicate block's bloom=%x", rbloom)
+		return
+	}
+
+	// Tre receipt Trie's root (R = (Tr [[H1, R1], ... [Hn, R1]]))
+	receiptSha := types.DeriveSha(receipts)
+	if receiptSha != header.ReceiptHash {
+		err = fmt.Errorf("invalid receipt root hash. received=%x calculated=%x", header.ReceiptHash, receiptSha)
+		return
+	}
+
+	// Commit state objects/accounts to a database batch and calculate
+	// the state root. The database is not modified if the root
+	// doesn't match.
+	root, batch := state.CommitBatch()
+	if header.Root != root {
+		err = fmt.Errorf("invalid merkle root: header=%x computed=%x", header.Root, root)
+		return
+	}
+	return batch, nil
+}
+
+// VarifyUncles verifies the given block's uncles and applies the Ethereum
+// consensus rules to the various block headers included; it will return an
+// error if any of the included uncle headers were invalid.
+func (sm *BlockValidator) VerifyUncles(block, parent *types.Block) error {
```

Comment:
A general question about this method: wouldn't it be worthwhile to rework this function so that it operates on only headers instead of blocks? Uncle verification should be imho doable on purely header chains too and not require the presence of block bodies (apart from the uncles being currently verified). 

I would imagine a function signature such as `VerifyUncles(header *types.Header, uncles []*types.Header)`. This would contain enough info to verify everything, without actually needing the presence of a lot of added fields.